TopMovies/API/CategoryNameCoder.swift

```diff
@@ -0,0 +1,36 @@
+//
+//  CategoryNameCoder.swift
+//  TopMovies
+//
+//  Created by Macbook Pro  on 15.02.2021.
+//
+
+import Foundation
+
+struct CategoryNameCoder {
+  static func encodeName(for target: MovieTarget.Target) -> String {
+    switch target {
+    case .nowPlaying(page: _):
+      return "Now Playing"
+    case .popular(page: _):
+      return "Popular"
+    case .topRated(page: _):
+      return "Top Rated"
+    case .upcoming(page: _):
+      return "Upcoming"
+    }
+  }
+  static func decodeTarget(from name: String) -> MovieTarget.Target? {
+    switch name {
+    case "Now Playing":
+      return .nowPlaying()
+    case "Popular":
+      return .popular()
+    case "Top Rated":
+      return .topRated()
+    case "Upcoming":
+      return .upcoming()
+    default: return nil
```

Comment:
I still don't understand why we need that NameCoder at all? You don't need to store `String` in the state, you can always store enum type over there.

The only place where you need String is Props. So you can simply create the following enum:
```swift
enum MovieCategory {
     case nowPlaying = "Now Playing"
     case popular = "Popular"
     case topRated = "Top Rated"
     case upcoming = "Upcoming"
}
```
And in props get the `rawValue` of it

```swift
self.categoryName = category.rawValue
```

A `MovieTarget.Target` (I'd prefer to change the name since it sounds creepy) can be a pure enum with page as associated value and you can always map one enum into another:
```swift
extension MovieTarget.RequestType { // Decide name on your own, I'm not fully aware of the context
  init(category: MovieCategory, page: Int) {
    switch category {
       case .nowPlaying: self = .nowPlaying(page: page)
       case .popular: self = .popular(page: page)
       case .topRated: self = .topRated(page: page)
       case .upcoming: self = .upcoming(page: page)
    }
  }
}
```